Improved README file (#57)
AbolfaZlRezaEe/NMock: README.md

> - [:collision: What is NMock?](#collision-what-is-nmock)

Reviewer: Actually, the purpose of designing the NMock banner was meant to be on the top of the README.md file. It's good to see an overview of the application at the start and then start to read the documentation

Author: ```suggestion
![NMock Banner](https://user-images.githubusercontent.com/73066290/198220049-159f1118-b181-415d-bc61-c9a0d7b1bf87.png)

- [:collision: What is NMock?](#collision-what-is-nmock)
```

  - [Data source](#data-source)
  - [Repositories](#repositories)
- [:arrow\_left::arrow\_right: Mock Import/Export file structure](#arrow_leftarrow_right-mock-importexport-file-structure)
  - [Export](#export)
  - [Import](#import)
- [:bookmark\_tabs: Helpful tips](#bookmark_tabs-helpful-tips)

## :collision: What is NMock?
Ever wondered how Android developers test map features and functionalities in their projects? Do they ride their cars or bikes to do so, or do they walk in the streets? The answer is no. 

There are applications in Android called Mock applications. They help us generate fake locations and trips to test map features. 

**NMock** is a tool that helps you generate fake locations and trips to test the map feature in your product. It enables you to share them with others as well. We've built this application with the help of the following products and services:
- **[Google Map SDK](https://developers.google.com/maps/documentation)**
-  **[Kotlin Language](https://kotlinlang.org/)** 
- **[Kotlin Coroutines](https://kotlinlang.org/docs/coroutines-overview.html)** 
- **[Dagger-Hilt Dependency Injection](https://dagger.dev/hilt/)**. 

> ![NMock Banner](https://user-images.githubusercontent.com/73066290/198220049-159f1118-b181-415d-bc61-c9a0d7b1bf87.png)

Author: ```suggestion
```

> This document is a guide on how this application works.

Reviewer: I think now we know that this README wanna explain the application in detail. so this does not necessarily need to write.

Author: ```suggestion
```

- Retrieve routing information for your trip.
  
Then you can store all this information in your local database. 

> **[Neshan API](https://platform.neshan.org/)** is one of the APIs we use to retrieve information. Check out their documentation as well.

### Repositories

We created three repositories to help us manage **ViewModel** requests. This section is critical because it creates and controls all processes. We receive data and transform it into a standard model that ViewModels can use and parse for **Views**.

The following image, shows the difference between the two of them:

![architecture models](https://user-images.githubusercontent.com/73066290/172200304-0e7baeb1-7ae0-462c-97a8-cea06299aee0.png)

> The code on the left represents the data structure we receive from the server, and the code on the right indicates the format we need for our view.

In ViewModels, we have two different approaches to accessing the views. Each ViewModel has two outputs that can be used by Views to manage their views and actions:
- The first one is a **Stateflow** which represents the state of the view. This flow contains all the information that the view needs to display (MVI approach)

> - The second one works with a **Sharedflow** and sends some actions that can either contain a message or be empty. View, then parse these actions and perform several actions, like displaying an error, for example. 

Reviewer: I think we use "action" in this sentence a lot. maybe it's good to write it like this:

`...View, then parse these actions and perform several works...`

Author: ```suggestion
- The second one works with a **Sharedflow** and sends some actions that can either contain a message or be empty. View, then parse these actions and perform several works, like displaying an error, for example. 
```

    }
```

One of the important actions that Sharedflows can do upon receiving an action from the ViewModel is to parse the action and undo something in view. For example, imagine that you sent a request to the server to retrieve data, but the server did not respond; now, we should send an action and hide or disable something that was shown or enabled before. The following image shows the flow:

![error architecture](https://user-images.githubusercontent.com/73066290/172200355-c851739b-7dbd-4ce2-a419-22a66a0b2bf3.png)

## :arrow_left::arrow_right: Mock Import/Export file structure

Sometimes you need to share your trip data with your friends or colleagues. We have implemented a feature allowing you to share your mocked data in JSON format easily.

### Export

> We store your mock trip data and its state in **JSON** format. We use the same format to export this data. The following JSON is how we structure this data to export:

Reviewer: The first sentence is wrong programmatically. We store mock trips in the database.

Author: ```suggestion
We store your mock trip data and its state in the database. We use the same format to export this data. The following JSON is how we structure this data to export:
```

}
```

|Name| Type | Description |
|--|--|--|
| **file_created_at** | String | Creation Date/Time of the exported JSON file
|**file_owner**|String|User UUID
|**version_code**|Number| Application version code
|**mock_information**|Object| Necessary mock data
|**route_information**|Object| Route information and line. **route_lines** is the most important field in this object. It is an array containing a list of positions (nodes) to draw lines on the map. Every array item has a **latitude** and **longitude** for its position.

### Import

As you see in the previous section, you can export and share your mock data with your friends and colleagues in JSON format. To import any data to your project, you can easily import this file using the import feature. ( We have plans in the near future to add other methods to import data )

We store this imported JSON data like standard mock data in a different database table. We add the following additional column to each row:
- fileCreatedAt
- fileOwner
- versionCode

We separated the imported data so we could manage them easily. These are the reasons that give us the ability to do so:

- We may have less imported data in the future.

> - We have a plan to implement a separate UI to display imported data.

Reviewer: Let me replace and improve this section with the following sentences:

- Maybe We have a different structure for the imported mocks in the future
- In the future, We wanna implement a different flow for showing imported mocks in UI and Logic too.

Author: ```suggestion
- Maybe We have a different structure for the imported mocks in the future
- In the future, We wanna implement a different flow for showing imported mocks in UI and Logic too.
```
